pkg/storage/etcd/etcd_helper.go

```diff
@@ -289,14 +289,20 @@ func (h *etcdHelper) decodeNodeList(nodes []*etcd.Node, slicePtr interface{}) er
 	for _, node := range nodes {
 		if node.Dir {
 			trace.Step("Decoding dir " + node.Key + " START")
-			if err := h.decodeNodeList(node.Nodes, slicePtr); err != nil {
+			if err := h.decodeNodeList(node.Nodes, filter, slicePtr); err != nil {
 				return err
 			}
 			trace.Step("Decoding dir " + node.Key + " END")
 			continue
 		}
 		if obj, found := h.getFromCache(node.ModifiedIndex); found {
-			v.Set(reflect.Append(v, reflect.ValueOf(obj).Elem()))
+			matches, err := filter(obj)
+			if err != nil {
```

Comment:
I question why we're making filters return errors. I think that everywhere you basically want to throw away the error, i.e., do `matches && err == nil`. I think allowing a filter function's error to abort an entire list (like here) is probably a mistake; imagine a wrongly typed object does somehow sneak in; this check may prevent ever listing the collection again, even to find the bad object to delete it.

So I tentatively think filter functions should continue to return only a bool. But I don't know your reason, so maybe you can convince me.